lib/tasks/gdpr_delete.rake

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+namespace :gdpr do
+  desc 'Deletes user accounts in accordance with GDPR. ' \
+       'Requires a text file with external user IDs (one per line) and ' \
+       'the corresponding consumer for deletion. Pass those via ' \
+       'TXT=/path/to/file.txt and CONSUMER=Name.'
+  task delete_users: :environment do
+    text_file_location = ENV['TXT']
+    consumer_name = ENV['CONSUMER']
+
+    if text_file_location.blank? || consumer_name.blank?
+      warn '--------------------------------------'
+      warn 'GDPR deletion tool to remove personal user data. USAGE:'
+      warn '--------------------------------------'
+      warn 'rake gdpr:delete_users TXT=./accounts.txt CONSUMER=openHPI'
+      exit 1
+    end
+
+    consumer = Consumer.where('lower(name) = ?', consumer_name.downcase).first
+
+    if consumer.blank?
+      warn '--------------------------------------'
+      warn "ERROR: The consumer '#{consumer_name}' was not found but is mandatory."
+      warn '--------------------------------------'
+      exit 2
+    end
+
+    file = begin
+      File.open(text_file_location, 'r')
+    rescue Errno => e
+      warn '--------------------------------------'
+      warn "ERROR: The file '#{text_file_location}' could not be opened: #{e.message}"
+      warn '--------------------------------------'
+      exit 3
+    end
+
+    unknown_user_ids = []
+    users_deleted = 0
+    file.each_line do |row|
```

Comment:
This shortcut comes in pretty handy when reading a file with data per line: `.readlines.map(&:chomp)`.
You'll get an array of IDs that can be iterated over.

Reply:
That's a nice suggestion as well. At first, I decided against this method, as it requires reading the complete file at once for the processing. Depending on the size, the other solution might be better for very large files. But since we don't expect _that_ large files, I think this aspect is not worth keeping. Therefore, I changed the implementation according to your suggestion.

Next, one could consider changing the way we interact with database objects. Performance-wise, the current implementation has many drawbacks: For each (valid) line in the text file, we have one `SELECT` and one `UPDATE` query. We could also change this to use query all user IDs at once (i.e., with an `ExternalUser.where()` clause) or even using `update_all`. I refrained from doing so, since I wanted to focus on error-free processing and catch all errors that occur. For the bulk-operation, I am currently not aware of a suitable mechanism to identify those user IDs that were not used to fetch a record.